## Notebook Overview

### 0.1.1 Project Overview

This notebook focuses on the crucial initial steps of a machine learning project for credit risk prediction. Specifically, we will load the raw Home Credit dataset, perform essential data cleaning and memory optimization, and save the preprocessed data in a more efficient format (Parquet) for subsequent stages of the project.

### 0.1.2 Why is this Important?

Efficient data loading and preprocessing are fundamental to any successful machine learning project. By optimizing memory usage and ensuring data quality, we lay the groundwork for faster processing, improved model training, and more reliable results. This is particularly important in credit risk prediction where accurate and timely decisions are crucial for financial institutions.

### 0.1.3 Dataset Description

The Home Credit dataset contains information about loan applications from Home Credit Group. It includes a variety of features related to applicant demographics, financial history, and loan specifics. This data will be used to build models that predict the likelihood of loan defaults. The dataset is structured across several interconnected tables, offering a rich source of information for analysis.

### 0.1.4 Our Approach

We will leverage the Polars library for data loading and manipulation due to its efficiency in handling large datasets. Our approach includes:

1. **Data Loading:** Load the `application_train.csv` and `application_test.csv` files using Polars.
2. **Initial Exploration:** Examine the first few rows and data types to gain an initial understanding of the data.
3. **Lowercase Column Names:** Standardize column names by converting them to lowercase.
4. **Memory Optimization:** Reduce the dataset's memory footprint by converting data types to lower-precision alternatives without significant loss of information.
5. **Save Preprocessed Data:** Save the optimized DataFrames to Parquet files for efficient storage and retrieval in subsequent notebooks.

### 0.1.5 Additional Notes

This notebook assumes that the raw data files (`application_train.csv` and `application_test.csv`) are located in the `../data/raw/` directory. The preprocessed data will be saved in the `../data/processed/` directory.

In [2]:
import polars as pl
from retail_bank_risk.data_preprocessing_utils import reduce_memory_usage_pl

We then load the CSV files using Polars.


In [3]:
application_train = pl.read_csv("../data/raw/application_train.csv")
application_test = pl.read_csv("../data/raw/application_test.csv")

We've loaded the main data tables, `application_train` and `application_test`, which contains the core details of each loan application.

This information might be sufficient for building our initial models.

If we need more information to improve accuracy, we'll then explore the other tables and incorporate relevant data from them.

Next up we will check the first rows of our train and test datasets.


In [4]:
application_train.head()

shape: (5, 122)
┌────────────┬────────┬────────────┬───────────┬───┬───────────┬───────────┬───────────┬───────────┐
│ SK_ID_CURR ┆ TARGET ┆ NAME_CONTR ┆ CODE_GEND ┆ … ┆ AMT_REQ_C ┆ AMT_REQ_C ┆ AMT_REQ_C ┆ AMT_REQ_C │
│ ---        ┆ ---    ┆ ACT_TYPE   ┆ ER        ┆   ┆ REDIT_BUR ┆ REDIT_BUR ┆ REDIT_BUR ┆ REDIT_BUR │
│ i64        ┆ i64    ┆ ---        ┆ ---       ┆   ┆ EAU_WEEK  ┆ EAU_MON   ┆ EAU_QRT   ┆ EAU_YEAR  │
│            ┆        ┆ str        ┆ str       ┆   ┆ ---       ┆ ---       ┆ ---       ┆ ---       │
│            ┆        ┆            ┆           ┆   ┆ f64       ┆ f64       ┆ f64       ┆ f64       │
╞════════════╪════════╪════════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡
│ 100002     ┆ 1      ┆ Cash loans ┆ M         ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 1.0       │
│ 100003     ┆ 0      ┆ Cash loans ┆ F         ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 0.0       │
│ 100004     ┆ 0      ┆ Revolving  ┆ M         ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 0.0       │
│            ┆        ┆ loans      ┆           ┆   ┆           ┆           ┆           ┆           │
│ 100006     ┆ 0      ┆ Cash loans ┆ F         ┆ … ┆ null      ┆ null      ┆ null      ┆ null      │
│ 100007     ┆ 0      ┆ Cash loans ┆ M         ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 0.0       │
└────────────┴────────┴────────────┴───────────┴───┴───────────┴───────────┴───────────┴───────────┘

In [5]:
application_test.head()

shape: (5, 121)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ SK_ID_CUR ┆ NAME_CONT ┆ CODE_GEND ┆ FLAG_OWN_ ┆ … ┆ AMT_REQ_C ┆ AMT_REQ_C ┆ AMT_REQ_C ┆ AMT_REQ_ │
│ R         ┆ RACT_TYPE ┆ ER        ┆ CAR       ┆   ┆ REDIT_BUR ┆ REDIT_BUR ┆ REDIT_BUR ┆ CREDIT_B │
│ ---       ┆ ---       ┆ ---       ┆ ---       ┆   ┆ EAU_WEEK  ┆ EAU_MON   ┆ EAU_QRT   ┆ UREAU_YE │
│ i64       ┆ str       ┆ str       ┆ str       ┆   ┆ ---       ┆ ---       ┆ ---       ┆ AR       │
│           ┆           ┆           ┆           ┆   ┆ f64       ┆ f64       ┆ f64       ┆ ---      │
│           ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆ f64      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 100001    ┆ Cash      ┆ F         ┆ N         ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 0.0      │
│           ┆ loans     ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 100005    ┆ Cash      ┆ M         ┆ N         ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 3.0      │
│           ┆ loans     ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 100013    ┆ Cash      ┆ M         ┆ Y         ┆ … ┆ 0.0       ┆ 0.0       ┆ 1.0       ┆ 4.0      │
│           ┆ loans     ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 100028    ┆ Cash      ┆ F         ┆ N         ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 3.0      │
│           ┆ loans     ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 100038    ┆ Cash      ┆ M         ┆ Y         ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │
│           ┆ loans     ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

We can see that there are many features. As expected, the test dataset does not include the target variable.

First lets lowercase the columns names and then lets check the exact number of rows and features we do have here.


In [6]:
application_train = application_train.select(
    [pl.col(name).alias(name.lower()) for name in application_train.columns]
)
application_test = application_test.select(
    [pl.col(name).alias(name.lower()) for name in application_test.columns]
)

In [7]:
print("Number of Rows and Columns in Application Train:", application_train.shape)
print("Number of Rows and Columns in Application Test:", application_test.shape)

Number of Rows and Columns in Application Train: (307511, 122)
Number of Rows and Columns in Application Test: (48744, 121)


Next up, before proceeding, we will reduce the dataset's memory footprint by converting the data types to lower-precision alternatives. This will lead to faster processing times.

In [8]:
application_train = reduce_memory_usage_pl(application_train)
application_test = reduce_memory_usage_pl(application_test)

Size before memory reduction: 289.61 MB
Initial data types: Counter({Float64: 65, Int64: 41, String: 16})
Size after memory reduction: 113.52 MB
Final data types: Counter({Float32: 65, Int8: 35, Categorical(ordering='physical'): 16, Int32: 4, Int16: 2})
Size before memory reduction: 45.51 MB
Initial data types: Counter({Float64: 65, Int64: 40, String: 16})
Size after memory reduction: 17.94 MB
Final data types: Counter({Float32: 65, Int8: 34, Categorical(ordering='physical'): 16, Int32: 4, Int16: 2})


We converted the data types of the `application_train` and `application_test` DataFrames to lower-precision alternatives.

This reduced the memory footprint of the training data from 289.61 MB to 113.52 MB and that of the test data from 45.51 MB to 17.94 MB.

That being said, let's confirm our features' data types.

In [9]:
application_train.head()

shape: (5, 122)
┌────────────┬────────┬────────────┬───────────┬───┬───────────┬───────────┬───────────┬───────────┐
│ sk_id_curr ┆ target ┆ name_contr ┆ code_gend ┆ … ┆ amt_req_c ┆ amt_req_c ┆ amt_req_c ┆ amt_req_c │
│ ---        ┆ ---    ┆ act_type   ┆ er        ┆   ┆ redit_bur ┆ redit_bur ┆ redit_bur ┆ redit_bur │
│ i32        ┆ i8     ┆ ---        ┆ ---       ┆   ┆ eau_week  ┆ eau_mon   ┆ eau_qrt   ┆ eau_year  │
│            ┆        ┆ cat        ┆ cat       ┆   ┆ ---       ┆ ---       ┆ ---       ┆ ---       │
│            ┆        ┆            ┆           ┆   ┆ f32       ┆ f32       ┆ f32       ┆ f32       │
╞════════════╪════════╪════════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡
│ 100002     ┆ 1      ┆ Cash loans ┆ M         ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 1.0       │
│ 100003     ┆ 0      ┆ Cash loans ┆ F         ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 0.0       │
│ 100004     ┆ 0      ┆ Revolving  ┆ M         ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 0.0       │
│            ┆        ┆ loans      ┆           ┆   ┆           ┆           ┆           ┆           │
│ 100006     ┆ 0      ┆ Cash loans ┆ F         ┆ … ┆ null      ┆ null      ┆ null      ┆ null      │
│ 100007     ┆ 0      ┆ Cash loans ┆ M         ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 0.0       │
└────────────┴────────┴────────────┴───────────┴───┴───────────┴───────────┴───────────┴───────────┘

We can see that the initial data exploration reveals that several features, despite being represented numerically, are inherently categorical.

These include `region_rating_client`, `region_rating_client_w_city`, `reg_city_not_work_city`, and `target`.

To ensure accurate analysis and modeling, we need to explicitly convert these features to the appropriate categorical data type.

In [14]:
categorical_cols_train = [
    "region_rating_client",
    "region_rating_client_w_city",
    "reg_city_not_work_city",
    "target",
]

categorical_cols_test = [
    "region_rating_client",
    "region_rating_client_w_city",
    "reg_city_not_work_city",
]

application_train = application_train.with_columns(
    [
        pl.col(col).round(0).cast(pl.Int32).cast(pl.Utf8).cast(pl.Categorical)
        for col in categorical_cols_train
    ]
)

application_test = application_test.with_columns(
    [
        pl.col(col).round(0).cast(pl.Int32).cast(pl.Utf8).cast(pl.Categorical)
        for col in categorical_cols_test
    ]
)

Next, we will save the preprocessed DataFrames (`application_train` and `application_test`) to Parquet files (`application_train_preprocessed.parquet` and `application_test_preprocessed.parquet`) for efficient storage and retrieval.

These files will be used in the next notebook (`02_feature_engineering.ipynb`) to perform feature engineering and further data preparation. 

In [10]:
application_train.write_parquet(
    "../data/processed/application_train_preprocessed.parquet"
)
application_test.write_parquet(
    "../data/processed/application_test_preprocessed.parquet"
)